### DISCUSSION TITLE
Best way to `SET` in Postgres?

### DISCUSSION THREAD

**Discussion by rustprooflabs** [psycopg/psycopg#79] [Q&A] [answered]
I am working on a migration path for an internal app to psycopg3 and am not sure if what I am encountering requires reworking my approach or if this is something that can be fixed. 

#### Versions

Python 3.9.5, psycopg3 beta 1 via PyPI.  Postgres 12.8.

#### Existing code

In psycopg2 I have used code such as the following.  The `account_id` is used along with Postgres' Row Level Security (RLS).  This is also how I have implemented a timezone setting.

```python
cur.execute('SET SESSION my.account_id TO %s;', [account_id])
cur.execute('SET TIMEZONE = %s;', [timezone])
```

Attempting the same in psycopg3 results in a `SyntaxError`.

```
psycopg.errors.SyntaxError: syntax error at or near "$1"
LINE 1: SET SESSION my.account_id TO $1;
```

Same basic error with the timezone, I tried using a dict instead of the simple parameter with the same basic error.

Should this be possible?  If not...

#### Possible workaround

The workaround I have found is to set these at the connection level.

```python
conn_options = f'-c my.account_id={account_id} -c timezone={timezone}'
try:
    conn = psycopg.connect(config.DATABASE_STRING, options=conn_options)
Except ...
```

At a glance this will work.  The catch is I want to use this with the connection pool and setting these at the connection level would mean I need to one pool per account.  I imagine this wouldn't be that difficult to implement in-app, but would diminish the overall effectiveness of the pool.

**Comment by dvarrazzo** [comment, marked as answer]
Yes, this is a known difference from using server-side parameters. it is [documented here](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#server-side-binding) .

There are two workarounds: one is to use the [`set_config()` function](https://www.postgresql.org/docs/13/functions-admin.html#FUNCTIONS-ADMIN-SET-TABLE):

```python
cur.execute("SELECT set_session('my.account_id', %s)", [account_id])
```

The other is to use client-side binding via the [`psycopg.sql` module](https://www.psycopg.org/psycopg3/docs/api/sql.html):
```python
cur.execute(sql.SQL('SET SESSION my.account_id TO {};').format(account_id))
```

unlike psycopg2 a `sql.Literal` is not needed in the second case.

The first variant is less familiar for psycopg2 users but is probably more efficient.

**Reply by rustprooflabs** [reply]
Doh, and you even have a timezone example in there.  Sorry for not reading closer!

**Reply by dvarrazzo** [reply]
No, most definitely that docs section needs expansion and prominence.

Also the first part is wrong, psycopg now rarely throws errors on wrong data types. There are still a few but not that jsonb one.

Docs love needed there.
